util/ipgen/clkmgr_gen.py

```diff
@@ -0,0 +1,79 @@
+# Copyright lowRISC contributors (OpenTitan project).
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Provides support functions for clkmgr ipgen."""
+
+import math
+from typing import List, Tuple
+from collections import OrderedDict
+
+from basegen.typing import ConfigT, ParamsT
+from topgen.clocks import Clocks, ClockSignal
+from topgen.lib import find_module
+
+def get_clkmgr_params(top: ConfigT) -> ParamsT:
+    """Gets parameters for clkmgr ipgen from top config."""
+    clocks = top["clocks"]
+    assert isinstance(clocks, Clocks)
+
+    typed_clocks = clocks.typed_clocks()
+    hint_names = typed_clocks.hint_names()
+
+    typed_clks = OrderedDict({
+        ty: {
+            nm: {
+                "src_name": sig.src.name,
+                "endpoint_ip": sig.endpoints[0][0]
+            }
+            for nm, sig in mp.items() if isinstance(sig, ClockSignal)
+        }
+        for ty, mp in typed_clocks._asdict().items() if isinstance(mp, dict)
```

Comment:
I really like this dict comprehension: it's very clean!

Reply:
Thanks, but I can't take credit for it: I didn't change the code when I moved it from topgen.py.